File: tree/tree1_lc144.py
"""
    二叉树的前序遍历
    (中左右)
"""

# Definition for a binary tree node.
from unittest import result


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right
class Solution:

    def preorderTraversal(self, root):

        # 递归法(思路二：分解问题)
        # 一棵二叉树的前序遍历结果 = 根节点 + 左子树的前序遍历结果 + 右子树的前序遍历结果。
        # 见参考链接

        # 递归法(思路一：遍历)
        result = []
        def traversal(root):
            if not root:    # 递归结束条件：遇到空节点就返回
                return
            result.append(root.val)     # 中
            traversal(root.left)        # 左
            traversal(root.right)       # 右
        traversal(root)
        return result

        # # 迭代法
        # result = []
        # if not root:
        #     return result
        # stack = [root]
        # while stack:
        #     cur = stack.pop()
        #     if cur:
        #         if cur.right:
        #             stack.append(cur.right)       # 右
        #         if cur.left:
        #             stack.append(cur.left)        # 左
        #         stack.append(cur)                 # 中
        #         stack.append(None)
        #     else:
        #         cur = stack.pop()
        #         result.append(cur.val)
        # return result

# 构建一棵树
#       5
#   4       6
# 1   2   7   8
node1, node2, node3, node4 = TreeNode(5), TreeNode(4), TreeNode(6), TreeNode(1)
node5, node6, node7 = TreeNode(2), TreeNode(7), TreeNode(8)

node1.left, node1.right = node2, node3
node2.left, node2.right = node4, node5
node3.left, node3.right = node6, node7

aaa = Solution()
bbb = aaa.preorderTraversal(node1)
print(bbb)      # [5, 4, 1, 2, 6, 7, 8]

File: cpp/tree/tree1_lc144.cpp
/*
    二叉树的前序遍历
    (中左右)
*/
#include <iostream>
#include <vector>
#include <stack>
using namespace std;

// Definition for a binary tree node.
struct TreeNode
{
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

// // 递归法
// class Solution
// {
// public:
//     void traversal(TreeNode *tree, vector<int> &result)
//     {
//         if (tree == NULL)
//             return;
//         result.push_back(tree->val);
//         traversal(tree->left, result);
//         traversal(tree->right, result);
//     }

//     vector<int> preorderTraversal(TreeNode *root)
//     {
//         vector<int> result;
//         traversal(root, result);
//         return result;
//     }
// };

// 迭代法
class Solution
{
public:
    vector<int> preorderTraversal(TreeNode *root)
    {
        vector<int> result;
        stack<TreeNode *> st;
        if (root != NULL)
        {
            st.push(root);
        }
        while (!st.empty())
        {
            TreeNode *node = st.top();
            if (node != NULL)
            {
                st.pop();
                if (node->right) st.push(node->right); // 右
                if (node->left) st.push(node->left); // 左
                st.push(node);           // 中
                st.push(NULL);
            }
            else
            {
                st.pop();
                node = st.top();
                st.pop();
                result.push_back(node->val);
            }
        }
        return result;
    }
};

void printArr(vector<int> &arr)
{
    for (int i = 0; i < arr.size(); i++)
    {
        cout << arr[i] << " ";
    }
    cout << endl;
}

int main()
{
    TreeNode *node7 = new TreeNode(8);
    TreeNode *node6 = new TreeNode(7);
    TreeNode *node5 = new TreeNode(2);
    TreeNode *node4 = new TreeNode(1);
    TreeNode *node3 = new TreeNode(6, node6, node7);
    TreeNode *node2 = new TreeNode(4, node4, node5);
    TreeNode *node1 = new TreeNode(5, node2, node3);
    Solution *obj = new Solution();
    vector<int> result = obj->preorderTraversal(node1);
    printArr(result); // 5 4 1 2 6 7 8
    return 0;
}